tidy3d/components/geometry.py

```diff
@@ -3105,8 +3108,149 @@ def _surface_area(self, bounds: Bound) -> float:
         return area
 
 
+class CustomSurfaceMeshGeometry(Geometry, ABC):
+    """Custom surface geometry given by a triangle mesh, as in the STL file format.
+
+    Example
+    -------
+    >>> import stl
+    >>> stl_mesh = stl.mesh.Mesh(np.zeros(4), dtype=stl.mesh.Mesh.dtype)
+    >>> # tetrahedron
+    >>> stl_mesh.vectors = [
+    >>>     [[1,0,0],[0,1,0],[0,0,1]],
+    >>>     [[0,0,0],[0,0,1],[0,1,0]],
+    >>>     [[0,0,0],[1,0,0],[0,0,1]],
+    >>>     [[0,0,0],[0,1,0],[1,0,0]]
+    >>> ]
+    >>> stl_geom = CustomSurfaceMeshGeometry.from_stl(stl_mesh)
+    """
+
+    mesh_dataset: Optional[SurfaceMeshDataset] = pydantic.Field(
```

Comment:
Is there a reason we need a `SurfaceMeshDataset`? Why not just put the `SurfaceMeshDataArray` directly here? For the other components, eg. `CustomMedium`, a dataset is needed because one data array per component (eps_ii) is required, but here I think we don't need that. Unless you anticipate needing it later?

Reply:
I think it could be needed later if we support other formats besides STL. I don't see too much harm in leaving it this way for now. It also might make handling the case where the data fails to load from json a bit cleaner